core/directory.go

```diff
@@ -259,11 +265,17 @@ func (dir *Directory) WithDirectory(ctx context.Context, subdir string, src *Dir
 		return nil, err
 	}
 
+	var perm *fs.FileMode = nil
+	if permissions != 0 {
+		perm = &permissions
+	}
+
 	st = st.File(llb.Copy(srcSt, srcPayload.Dir, path.Join(destPayload.Dir, subdir), &llb.CopyInfo{
 		CreateDestPath:      true,
 		CopyDirContentsOnly: true,
 		IncludePatterns:     filter.Include,
 		ExcludePatterns:     filter.Exclude,
+		Mode:                perm,
```

Comment:
Does this only affect the root directory?

Reply:
I'm honestly not sure. I'll add a test to check.

Reply:
Turns out the answer is no... it impacts all sub directories and files.

Reply:
Hm, that's unfortunate, and feels kind of surprising. Thanks for looking into it.

I wonder if we should just leave this API out instead for now. I can't think of many scenarios where I'd want to `chmod` an entire directory recursively to the same permissions. I guess `777` would make sense, but not much else.

cc @sipsma @shykes @aluzzardi for a second opinion :)

Reply:
Yeah, I agree about it being surprising. I'm definitely open to pulling it from `Container|Directory.withDirectory()` to reduce the element of surprise.

Reply:
Agree recursive chmod would be surprising behavior and that we can just drop this from `withDirectory` for now to unblock.

If we did want the API to work such that it only applied to the directory and not contents, I wonder if you could do that by first doing a normal copy call without mode set and then running another copy that uses either include or exclude patterns such that the mode only applies to the directory. But that seems hacky and a lot of hoops to jump through; we can also look into upstream changes if there's no other existing options and we have a strong need for it.